.github/workflows/ghcr.yml

```diff
@@ -0,0 +1,45 @@
+name: Build and publish multi-arch container images
+
+on:
+  push:
+    branches: [ main ]
+  workflow_dispatch:
+    inputs:
+      reason:
+        description: 'Why manual trigger?'
+        required: false
+        default: ''
+
+jobs:
+  ghcr_build_and_push:
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push'
```

Comment:
Does this mean the workflow will build an image and push it to `ghcr.io` on every pushed commit? Is it possible to do this only when the `Dockerfile` is changed?

Reply:
I have no idea about such trigger condition, I'll try to find it is possible to do that later.

Reply:
We'll want to commit every time the _code_ changes. But we definitely want to limit it to the `main` branch, or we'll be overwriting it with our feature branches

Reply:
I find some tricks to realize it:
1. detect all Dockerfile are changed. this method will trigger the workflow when any Dockerfile changed.
```
on:
  push:
    branches: [ main ]
    paths:
      - '**/Dockerfile'   // here.
```
2. compare the recent commits to confirm it is need to build. this method need to checkout multi commits(the depth is 2 when we decide to compare the last two commits) and may skip building when multi commits be merged, sth like this:
```
git diff --name-only ${{ github.event.before }} ${{ github.sha }} | grep "$dockerfile_dir/Dockerfile"
if [ $? -eq 0 ]; then
  need_build=true 
  break
fi
```
btw, we also can combine two methods.

how do you think about this?  or do you have any other solutions?

Reply:
@rbren do you think we need to re-build images on every main branch change, or just `Dockerfile` change? If you think we should rebuild on every git commits, we probably don't need to do this trick?